sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt

```diff
@@ -291,96 +317,124 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
+        this.surface?.release()
+      }
+      this.surface = surface
+    }
+    this.width = width
+    this.height = height
+    // we clean only Mapbox events to avoid outdated runnables associated with previous EGL context
+    renderEventQueueLock.withLock {
+      // using iterator to avoid concurrent modification exception
+      val iterator = renderEventQueue.iterator()
+      while (iterator.hasNext()) {
+        val next = iterator.next()
+        if (next.eventType == EventType.SDK) {
+          iterator.remove()
+        }
+      }
+    }
+    // we do not want to clear render events scheduled by user
+    renderHandlerThread.clearMessageQueue(clearAll = false)
+    prepareRenderFrame(creatingSurface = true)
+  }
+
   @UiThread
   fun onSurfaceCreated(surface: Surface, width: Int, height: Int) {
     lock.withLock {
       renderHandlerThread.post {
-        if (this.surface != surface) {
-          releaseEgl()
-          this.surface?.release()
-          this.surface = surface
-        }
-        this.width = width
-        this.height = height
-        shouldExit = false
-        eventQueue.clear()
-        renderEventQueue.clear()
-        snapshotQueue.clear()
-        renderHandlerThread.clearMessageQueue()
-        prepareRenderFrame(creatingSurface = true)
+        processAndroidSurface(surface, width, height)
       }
       createCondition.await()
     }
   }
 
   @AnyThread
   fun setMaximumFps(fps: Int) {
-    renderTimeNs.set(ONE_SECOND_NS / fps)
+    renderTimeNs = ONE_SECOND_NS / fps
   }
 
   @WorkerThread
   override fun doFrame(frameTimeNanos: Long) {
-    try {
-      awaitingNextVsync.set(false)
-      if (eglPrepared && !paused && !shouldExit) {
-        draw()
-      }
-    } finally {
-      // regardless if we did drawing or not execute all non gl relative events
-      drainQueue(eventQueue)
+    // it makes sense to draw not only when EGL config is prepared but when native renderer is created
+    if (renderThreadPrepared && !paused) {
+      draw()
     }
-  }
-
-  // MapRenderer delegate methods
-
-  @AnyThread
-  fun requestRender() {
-    postPrepareRenderFrame()
+    awaitingNextVsync = false
   }
 
   @AnyThread
-  fun queueRenderEvent(runnable: Runnable) {
-    renderEventQueue.add(runnable)
-    postPrepareRenderFrame()
+  fun queueRenderEvent(renderEvent: RenderEvent) {
+    if (renderEvent.needRender) {
+      renderEvent.runnable?.let {
+        renderEventQueueLock.withLock {
+          renderEventQueue.add(renderEvent)
+        }
+      }
+      // in case of native Mapbox events we schedule render event only when we're fully ready for render
+      // TODO ideally to be fixed in core
+      if (renderEvent.eventType == EventType.SDK) {
+        if (renderThreadPrepared) {
+          postPrepareRenderFrame()
+        }
+      } else {
+        postPrepareRenderFrame()
+      }
+    } else {
+      // no sense in scheduling non-render SDK tasks if thread is not fully ready for render
```

Comment:
But render SDK events are NOT added also, right? Maybe simplify / squash both comments to some short and easy one?

Reply:
didn't get it fully - although condition is the same but in one case it's `postNonRenderEvent(renderEvent)`, in other `postPrepareRenderFrame()` - I'll need some lambda param then, not sure it's worth it

Reply:
Ok indeed